extensions/objects/templates/numeric-expression-editor.component.ts

```diff
@@ -19,104 +19,135 @@
-        let eventType = (
-          DeviceInfoService.isMobileUserAgent() &&
-          DeviceInfoService.hasTouchEvents()) ? 'focus' : 'change';
-        // We need the 'focus' event while using the on screen keyboard (only
-        // for touch-based devices) to capture input from user and the 'change'
-        // event while using the normal keyboard.
-        Guppy.event(eventType, (focusObj) => {
-          if (!focusObj.focused) {
-            ctrl.isCurrentAnswerValid();
-          }
-          var activeGuppyObject = (
-            GuppyInitializationService.findActiveGuppyObject());
-          if (activeGuppyObject !== undefined) {
-            ctrl.hasBeenTouched = true;
-            ctrl.currentValue = activeGuppyObject.guppyInstance.asciimath();
-            if (eventType === 'change') {
-              // Need to manually trigger the digest cycle to make any
-              // 'watchers' aware of changes in answer.
-              $scope.$apply();
-            }
-          }
-        });
-        if (eventType !== 'focus') {
-          Guppy.event('focus', (focusObj) => {
-            if (!focusObj.focused) {
-              ctrl.isCurrentAnswerValid();
-            }
-          });
+      });
+      Guppy.event('focus', (focusObj) => {
+        if (!focusObj.focused) {
+          this.isCurrentAnswerValid();
         }
-      };
+      });
     }
-  ]
-});
+  }
+
+  isCurrentAnswerValid(): boolean {
+    if (this.currentValue === undefined) {
+      this.currentValue = '';
+    }
+    // Replacing abs symbol, '|x|', with text, 'abs(x)' since the symbol
+    // is not compatible with nerdamer or with the backend validations.
+    this.currentValue = this.mathInteractionsService.replaceAbsSymbolWithText(
+      this.currentValue);
+    var answerIsValid = this.mathInteractionsService.validateNumericExpression(
+      this.currentValue);
+    if (this.guppyInitializationService.findActiveGuppyObject() === undefined) {
+      // The warnings should only be displayed when the editor is inactive
+      // focus, i.e., the user is done typing.
+      this.warningText = this.mathInteractionsService.getWarningText();
+    } else {
+      this.warningText = '';
+    }
+    if (answerIsValid) {
+      // Explicitly inserting '*' signs wherever necessary.
+      this.currentValue = (
+        this.mathInteractionsService.insertMultiplicationSigns(
+          this.currentValue));
+      this.value = this.currentValue;
+      this.valueChanged.emit(this.value);
+    }
+    if (!this.hasBeenTouched) {
+      this.warningText = '';
+    }
+    this.eventBusGroup.emit(new ObjectFormValidityChangeEvent({
+      modalId: this.modalId,
+      value: !answerIsValid
+    }));
+    return answerIsValid;
+  }
+
+  showOSK(): void {
+    this.guppyInitializationService.setShowOSK(true);
+    GuppyInitializationService.interactionType = 'NumericExpressionInput';
+  }
+}
+
+angular.module('oppia').directive(
```

Comment:
Do we need the downgrade?

Reply:
Yup, we still $compile this component in object-editor.directive